runtime/lua/vim/lsp/folding_range.lua

```diff
@@ -0,0 +1,320 @@
+---@param bufnr integer
+local function schedule_foldupdate(bufnr)
+  if #(api.nvim_get_autocmds({
+    group = augroup_update,
+    buffer = bufnr,
+  })) == 0 then
+    api.nvim_create_autocmd('InsertLeave', {
+      group = augroup_update,
+      buffer = bufnr,
+      once = true,
+      callback = function()
+        foldupdate(bufnr)
+      end,
+    })
+  end
+end
+
+--- |lsp-handler| for the method `textDocument/foldingRange`
+--- Store folding range then update folds for a specific buffer.
+---@private
+---@param result lsp.FoldingRange[]?
+---@param ctx lsp.HandlerContext
+function M.on_folding_range(err, result, ctx)
+  if err then
+    log.error(err)
+    return
+  end
+
+  if not result then
+    return
+  end
+
+  local bufnr = assert(ctx.bufnr)
+  local bufstate = assert(bufstates[bufnr])
+
+  ---@type table<integer, [integer, nil | ">" | "<"]>
+  local row_level = {}
+  ---@type table<integer, string>
+  local row_text = {}
+  ---@type table<lsp.FoldingRangeKind, lsp.FoldingRange[]?>>
+  local kind_ranges = {}
+
+  for _, range in ipairs(result) do
+    local kind = range.kind
+    if kind then
+      ---@cast kind lsp.FoldingRangeKind
+      local ranges = kind_ranges[kind] or {}
+      ranges[#ranges + 1] = range
+      kind_ranges[kind] = ranges
+    end
+
+    local start_row = range.startLine
+    local end_row = range.endLine
+    -- Fold a single line is not supported.
+    if start_row ~= end_row then
+      row_text[start_row] = range.collapsedText
+      for row = start_row, end_row do
+        local level = row_level[row] or { 0 }
+        level[1] = level[1] + 1
+        row_level[row] = level
+      end
+      row_level[start_row][2] = '>'
+      row_level[end_row][2] = '<'
+    end
+  end
+
+  bufstate.row_level = row_level
+  bufstate.row_text = row_text
+  bufstate.kind_ranges = kind_ranges
+  bufstate.version = ctx.version
+
+  if api.nvim_get_mode().mode:match('^i') then
+    -- `foldUpdate()` is guarded in insert mode.
+    schedule_foldupdate(bufnr)
+  else
+    foldupdate(bufnr)
+  end
+end
+
+local augroup_refresh = vim.api.nvim_create_augroup('vim_lsp_folding_range/refresh', {})
+
+---@param bufnr integer
+---@param client_id integer The ID of the client whose server supports `foldingRange`.
+---@return integer? request_id
+local function refresh(bufnr, client_id)
+  ---@type lsp.FoldingRangeParams
+  local params = { textDocument = util.make_text_document_params(bufnr) }
+  local client = assert(vim.lsp.get_client_by_id(client_id))
+  local _, request_id = client:request(ms.textDocument_foldingRange, params, nil, bufnr)
+  return request_id
+end
+
+--- Setup event hooks for folding range.
+---@param bufnr integer
+---@return vim.lsp.folding_range.BufState
+local function setup(bufnr)
+  ---@type vim.lsp.folding_range.BufState
+  local bufstate = { row_level = {}, row_text = {}, kind_ranges = {} }
+  bufstates[bufnr] = bufstate
+
+  api.nvim_buf_attach(bufnr, false, {
```

Comment:
Need to also detach the buffer when all clients detach too.

Reply:
Is [this](https://github.com/ofseed/neovim/blob/e2a24c75c34a0c3adb6caf318d7a289f0c27c874/runtime/lua/vim/lsp/folding_range.lua#L172-L189) not enough?

Detaching the buffer should be unnecessary, because if this buffer has already been setup, then even if all clients detach at a certain moment (or any client has not yet attached), it is possible that it will be attached by new clients in the future.

Reply:
If there are no clients, then the buf attach callbacks should not be invoked.

Reply:
If you are referring to `on_bytes`, then it indeed will not be invoked, because when there are no clients, [here](https://github.com/ofseed/neovim/blob/e2a24c75c34a0c3adb6caf318d7a289f0c27c874/runtime/lua/vim/lsp/folding_range.lua#L172-L189) the row_levels will be cleared. This is intentional, because `row_level` being empty covers a wider range of possibilities.

Reply:
`on_bytes` will still be invoked until you detach from the buffer.

Reply:
Are you referring to directly removing the callback set by this buf attach? Then we would need to find a way to add it back, which would significantly increase complexity. I think it's sufficient to return after it finds the table is empty. Callbacks created by `buf_attach` will only be triggered on the buffer associated with a window whose `vim.wo.foldexpr = "v:lua.vim.lsp.foldexpr"`, so it makes sense in the vast majority of cases. I'm not quite sure what your suggested approach is.

Reply:
You reattach when you set up. You just need to return true in on_bytes to detach. It isn't that hard, it is just basic hygiene to clean up after all clients have detached.

Reply:
I think clearing bufstate is not an option. Because bufstate represents a buffer's window's `foldexpr` is `vim.lsp.foldexpr`, if it is cleared, next time `vim.lsp.foldexpr` is calculated, setup will be called again, and the callback will be created.

This is a design based on the callbacks and the bufstate has the same lifecycle. If the logic of the code is unclear, I can add more explanations later. However, it may be necessary to create an autocmd for `OptionSet` to clear bufstate and these callbacks when `foldexpr` is not `vim.lsp.foldexpr`.

Reply:
According to my understanding, you mean that for a buffer with `foldexpr` set to `vim.lsp.foldexpr`, it can be attached by LSP before or after (it's a goal), but during the time it is not attached (even if `foldexpr` is still `vim.lsp.foldexpr`), all callback executions need to be stopped.

This can be done through `LspAttach` and `LspDetach`, but I don't think it brings any real benefits. Especially during this time period, these callbacks return almost immediately